src/target/renesas.c

```diff
@@ -0,0 +1,938 @@
+	const bool code_flash = addr < RENESAS_CF_END;
+
+	/* Set Erasure Priority Mode */
+	target_mem_write16(t, RV40_FCPSR, RV40_FCPSR_ESUSPMD);
+
+	while (len) {
+		/* Set block start address*/
+		target_mem_write32(t, RV40_FSADDR, addr);
+
+		/* increment block address */
+		uint16_t block_size;
+		if (code_flash)
+			block_size = addr < RV40_CF_REGION0_SIZE ? RV40_CF_REGION0_BLOCK_SIZE : RV40_CF_REGION1_BLOCK_SIZE;
+		else
+			block_size = RV40_DF_BLOCK_SIZE;
+
+		addr += block_size;
+		len -= block_size;
+
+		/* Issue two part Block Erase commands */
+		target_mem_write8(t, RV40_CMD, RV40_CMD_BLOCK_ERASE);
+		target_mem_write8(t, RV40_CMD, RV40_CMD_FINAL);
+
+		/* according to reference manual the max erase time for a 32K block is around 1040ms
+		 * this is with a FCLK of 4MHz
+		 */
+		platform_timeout timeout;
+		platform_timeout_set(&timeout, 1100);
+
+		/* Wait until the operation has completed or timeout */
+		/* Read FRDY bit until it has been set to 1 indicating that the current  operation is complete.*/
+		while (!(target_mem_read32(t, RV40_FSTATR) & RV40_FSTATR_RDY)) {
+			if (target_check_error(t) || platform_timeout_is_expired(&timeout))
+				return -1;
+		}
+
+		if (renesas_rv40_error_check(t, RV40_FSTATR_ERSERR | RV40_FSTATR_ILGLERR))
+			return -1;
+	}
+
+	return 0;
+}
+
+static int renesas_rv40_flash_write(target_flash_s *f, target_addr_t dest, const void *src, size_t len)
+{
+	target *t = f->t;
+
+	/* code flash or data flash operation */
+	const bool code_flash = dest < RENESAS_CF_END;
+
+	/* write size for code flash / data flash */
+	const uint8_t write_size = code_flash ? RV40_CF_WRITE_SIZE : RV40_DF_WRITE_SIZE;
+
+	while (len) {
+		/* set block start address */
+		target_mem_write32(t, RV40_FSADDR, dest);
+
+		/* increment destination address */
+		dest += write_size;
+		len -= write_size;
+
+		/* issue two part Write commands */
+		target_mem_write8(t, RV40_CMD, RV40_CMD_PROGRAM);
+		target_mem_write8(t, RV40_CMD, (uint8_t)(write_size / 2U));
+
+		/* according to reference manual the data buffer full time for 2 bytes is 2 usec.
+		 * this is with a FCLK of 4MHz
+		 * a complete should take less than 1 msec.
+		 */
+		platform_timeout timeout;
+		platform_timeout_set(&timeout, 10);
+
+		/* write one chunk */
+		for (size_t i = 0U; i < (write_size / 2U); i++) {
+			/* copy data from source address to destination */
+			target_mem_write16(t, RV40_CMD, *(uint16_t *)src);
+
+			/* 2 bytes of data */
+			src += 2U;
+		}
+
+		/* issue write end command */
+		target_mem_write8(t, RV40_CMD, RV40_CMD_FINAL);
+
+		/* wait until the operation has completed or timeout */
+		/* read FRDY bit until it has been set to 1 indicating that the current operation is complete.*/
+		while (!(target_mem_read32(t, RV40_FSTATR) & RV40_FSTATR_RDY)) {
+			if (target_check_error(t) || platform_timeout_is_expired(&timeout))
+				return -1;
+		}
+	}
+
+	if (renesas_rv40_error_check(t, RV40_FSTATR_PRGERR | RV40_FSTATR_ILGLERR))
```

Comment:
This looks like another good target for return elision if we first merge the Flash return type PR before this one